security: use certs-dir for all certificates, redo certs commands

mberhault opened the pull request:
Part of #1674.

The `--certs-dir` flag replaces all the certs flags, with a naming
scheme used to determine what type each file is.
The user (needed for `client.<user>.{crt,key}`) comes from
`base.Config.User`.

Commit:
security: use certs-dir only for certificates/keys.

Part of #1674.

* use `--certs-dir` for all commands accepting certs
* accept old certs flags for `start` command only, but log deprecation warning
* rework certs creation logic
* make `--insecure` always default to false and independent of `--host`

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14703)
<!-- Reviewable:end -->

bdarnell commented:
I don't think we want to remove support for all the old flags - that will be disruptive to every existing secure deployment. I would keep the individual flags in place.

mberhault commented:
Yup, this is why it's not ready to look at yet, but I want to make sure everything works without the old flags, then use them as a fallback on the `start` command if we don't have a certs dir.

mberhault commented:
I still need to re-add the `cert/key/ca-cert` flags for the `start` command (the only one that gets them).
However, I will have a `these flags are deprecated, use --certs-dir instead` warning. We also need to figure out when to drop them. Ideally before 1.0.

mberhault commented:
old security flags re-added for `cockroach start` only. If any of them are specified, we attempt to use all of them.

mberhault commented:
SIGHUP-triggered reload is implemented in the last commit.

mberhault commented:
Actual rotation is in #14925, I figured this was getting big enough.

tamird commented:
Reviewed 5 of 14 files at r1, 1 of 2 files at r2, 1 of 13 files at r3, 4 of 11 files at r5, 5 of 9 files at r6, 2 of 2 files at r7.
Review status: 17 of 29 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/base/config.go, line 145 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-KhgaRyMKtjaqvvbDwmf:-KhgaRyMKtjaqvvbDwmg:bqlmhe7) ([raw file](https://github.com/cockroachdb/cockroach/blob/6960d3fdd3a8c3d01bd49bed523cbb068307932c/pkg/base/config.go#L145)):*
> ```Go
> 
> func didYouMeanInsecureError(err error) error {
> 	return fmt.Errorf("problem using security settings: %v, did you mean to use --insecure?", err)
> ```

`return errors.Wrap(err, ...)`

but looking at how this is used, i think it's likely to produce weird errors in non-CLI uses. Seems fragile to tie this library to CLI flags so directly.

---

*[pkg/base/config.go, line 149 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-KhgaVfaj5PsediwV2k0:-KhgaVfaj5PsediwV2k1:br9q6p3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6960d3fdd3a8c3d01bd49bed523cbb068307932c/pkg/base/config.go#L149)):*
> ```Go
> 
> // InitDefaults sets up the default values for a config.
> // This is also used in tests to reset global objects.
> ```

add an empty comment line above this to preserve the line break.

---

*[pkg/base/config.go, line 177 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-KhgackXpCuOEK1QehQy:-KhgackXpCuOEK1QehQz:b-1ybg08) ([raw file](https://github.com/cockroachdb/cockroach/blob/6960d3fdd3a8c3d01bd49bed523cbb068307932c/pkg/base/config.go#L177)):*
> ```Go
> 	if cfg.hasOldCertsFlags() {
> 		if cfg.SSLCert == "" || cfg.SSLCertKey == "" {
> 			return "", "", errors.New("client certificates use requires both --cert and --key")
> ```

can this error be more direct in telling the user what exactly is wrong?

---

*[pkg/security/certificate_loader.go, line 162 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-Khgc0FCBe8C-eBWOYEa:-Khgc0FCBe8C-eBWOYEb:bb7r9an) ([raw file](https://github.com/cockroachdb/cockroach/blob/6960d3fdd3a8c3d01bd49bed523cbb068307932c/pkg/security/certificate_loader.go#L162)):*
> ```Go
> 
> 	if !os.IsNotExist(err) {
> 		return errors.Errorf("could not stat certs directory %s: %v", cl.certsDir, err)
> ```

wrap

---

*[pkg/security/certificate_loader.go, line 166 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-Khgc12hGrEulTwh9co5:-Khgc12hGrEulTwh9co6:bb7r9an) ([raw file](https://github.com/cockroachdb/cockroach/blob/6960d3fdd3a8c3d01bd49bed523cbb068307932c/pkg/security/certificate_loader.go#L166)):*
> ```Go
> 
> 	if err := os.Mkdir(cl.certsDir, defaultCertsDirPerm); err != nil {
> 		return errors.Errorf("could not create certs directory %s: %v", cl.certsDir, err)
> ```

wrap

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: 17 of 29 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/base/config.go, line 145 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-KhgaRyMKtjaqvvbDwmf:-Khgh2NEeCGKyvQyNHhk:bzvny4) ([raw file](https://github.com/cockroachdb/cockroach/blob/6960d3fdd3a8c3d01bd49bed523cbb068307932c/pkg/base/config.go#L145)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

`return errors.Wrap(err, ...)`

but looking at how this is used, i think it's likely to produce weird errors in non-CLI uses. Seems fragile to tie this library to CLI flags so directly.
</blockquote></details>

Done.
This will be surfaced pretty quickly. The server loads configs right away, and clients use them to setup connections pretty quickly.

---

*[pkg/base/config.go, line 149 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-KhgaVfaj5PsediwV2k0:-KhghGJX6C0J-nDTFXxC:b-nmnudm) ([raw file](https://github.com/cockroachdb/cockroach/blob/6960d3fdd3a8c3d01bd49bed523cbb068307932c/pkg/base/config.go#L149)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

add an empty comment line above this to preserve the line break.
</blockquote></details>

You mean add some spacing in this comment? as in:
```
// InitDefaults sets up the default values for a config.
//
// This is also used in tests to reset global objects.
```

---

*[pkg/base/config.go, line 177 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-KhgackXpCuOEK1QehQy:-KhghSfN_GA3Y1w6QoNN:b-qn04cv) ([raw file](https://github.com/cockroachdb/cockroach/blob/6960d3fdd3a8c3d01bd49bed523cbb068307932c/pkg/base/config.go#L177)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

can this error be more direct in telling the user what exactly is wrong?
</blockquote></details>

Done, but those flags are being deprecated anyway.

---

*[pkg/security/certificate_loader.go, line 162 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-Khgc0FCBe8C-eBWOYEa:-KhggNqLHi6wUlhUVLRE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6960d3fdd3a8c3d01bd49bed523cbb068307932c/pkg/security/certificate_loader.go#L162)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

wrap
</blockquote></details>

Done.

---

*[pkg/security/certificate_loader.go, line 166 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-Khgc12hGrEulTwh9co5:-KhggOAPgS-pNe-QJCsQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6960d3fdd3a8c3d01bd49bed523cbb068307932c/pkg/security/certificate_loader.go#L166)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

wrap
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 14 files at r1, 1 of 2 files at r11, 5 of 13 files at r12, 6 of 11 files at r14, 7 of 9 files at r15, 5 of 5 files at r16.
Review status: all files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/base/config.go, line 139 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-KhrjWNrhlkIKJYiIakK:-KhrjWNrhlkIKJYiIakL:b61aq8h) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c2423ef3839e978c15ba68b86ddcd990c28754/pkg/base/config.go#L139)):*
> ```Go
> 
> 	log.Warning(context.Background(),
> 		"flags --ca-cert, --cert, and --key will soon be deprecated, please use --certs-dir",
> ```

They're deprecated now and will soon be removed. But I'm still not sure we want to be so aggressive about removing them. What harm does it do to keep them around? (I recognize that with 1.0 coming up we either remove them quickly or we'll be stuck with them forever)

---

*[pkg/cli/cert.go, line 33 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-KhrlGCd2kpgaH0s4bMR:-KhrlGCd2kpgaH0s4bMS:b-hdilti) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c2423ef3839e978c15ba68b86ddcd990c28754/pkg/cli/cert.go#L33)):*
> ```Go
> const defaultKeySize = 2048
> const defaultCALifetime = 5 * 265 * 24 * time.Hour // five years
> const defaultCertLifetime = 265 * 24 * time.Hour   // one year
> ```

s/265/365/g

A 5-year span will include one or possibly two leap years (and a one-year span might be a leap year), so these certs could expire a day or two earlier than expected. Maybe we should just assume a 366-day year (or do the right date math) so that if someone sets a reminder on their calendar 1 or 5 years out they won't be surprised by an expired cert.

---

*[pkg/cli/cert.go, line 46 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-KhrmqJ-lgxaFyMQUDah:-KhrmqJ0ubREYsWd8eai:b4x0n4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c2423ef3839e978c15ba68b86ddcd990c28754/pkg/cli/cert.go#L46)):*
> ```Go
> Generate a CA certificate "<certs-dir>/ca.crt" and CA key "<ca-key>".
> The certs directory is created if it does not exist.
> The CA key is created if it does not exist.
> ```

And if the CA key exists, what happens? Do we generate a new cert with the old key? (this would defeat the "select the ca cert based on which works with the key) approach mentioned below)

---

*[pkg/cli/cert.go, line 71 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-KhrmPYfbbKq77MtcMoL:-KhrmPYfbbKq77MtcMoM:b-vr85g7) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c2423ef3839e978c15ba68b86ddcd990c28754/pkg/cli/cert.go#L71)):*
> ```Go
> 	Long: `
> Generate a node certificate "<certs-dir>/node.crt" and key "<certs-dir>/node.key".
> Any existing files are overwritten.
> ```

Maybe we should require a `--force` flag to overwrite existing files.

---

*[pkg/cli/cert.go, line 84 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-KhrmZYmL2F8Mz4kS3TC:-KhrmZYmL2F8Mz4kS3TD:bgongl9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c2423ef3839e978c15ba68b86ddcd990c28754/pkg/cli/cert.go#L84)):*
> ```Go
> // TODO(marc): there is currently no way to specify which CA cert to use if more
> // than one is present. We shoult try to load each certificate along with the key
> // and pick the one that works. That way, the key specifies the certificate.
> ```

Some crypto libraries support simply concatenating the certificate to the private key file. If Go's does too, then we could store a second copy of the cert in the ca.key file and then use that one to avoid ambiguity.

---

*[pkg/cli/interactive_tests/test_old_secure.tcl, line 17 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-KhrndHsuAOXr55EZ-lS:-KhrndHtIM7poFywrbCd:b1c65gk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c2423ef3839e978c15ba68b86ddcd990c28754/pkg/cli/interactive_tests/test_old_secure.tcl#L17)):*
> ```Tcl
> 
> proc stop_secure_server {argv certs_dir} {
>     system "set -e; if kill -CONT `cat server_pid`; then $argv quit --certs-dir=$certs_dir || true & sleep 1; kill -9 `cat server_pid` || true; else $argv quit --certs-dir=$certs_dir || true; fi"
> ```

We handle `kill` (with the default `SIGTERM`) and `cockroach quit` the same way, so you can just kill the process instead of using `kill -CONT` to see whether it exists and then using `quit`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-:-KhrqtvD4rxkNPvDdRth:bnfp4nl)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 14 files at r1, 1 of 2 files at r11, 1 of 13 files at r12, 6 of 11 files at r14, 7 of 9 files at r15, 5 of 5 files at r16.
Review status: all files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[pkg/base/config.go, line 130 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-Khryyu7RMjj3zZPZEQZ:-Khryyu7RMjj3zZPZEQ_:bn0esd4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c2423ef3839e978c15ba68b86ddcd990c28754/pkg/base/config.go#L130)):*
> ```Go
> 
> // hasOldCertsFlags returns true if we have old certs flags and they are allowed.
> func (cfg *Config) hasOldCertsFlags() bool {
> ```

nit, but you call it `CertFlags` in the environment variable and `CertsFlags` here. You should use the same pluralization everywhere

---

*[pkg/cli/cert.go, line 33 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-KhrlGCd2kpgaH0s4bMR:-Khs-EwE6YqxKP5w3WLM:b-ydp5dr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c2423ef3839e978c15ba68b86ddcd990c28754/pkg/cli/cert.go#L33)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

s/265/365/g

A 5-year span will include one or possibly two leap years (and a one-year span might be a leap year), so these certs could expire a day or two earlier than expected. Maybe we should just assume a 366-day year (or do the right date math) so that if someone sets a reminder on their calendar 1 or 5 years out they won't be surprised by an expired cert.
</blockquote></details>

Also, I'd again request that we set these defaults higher such that only people who opt in to rotation ever have to deal with it. 500 years / 100 years, perhaps? Or no expiration?

---

*[pkg/cli/cert.go, line 124 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-Khs1PFLjjnGyMCoBpGK:-Khs1PFLjjnGyMCoBpGL:b-367svf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c2423ef3839e978c15ba68b86ddcd990c28754/pkg/cli/cert.go#L124)):*
> ```Go
> 	var username string
> 	if username, err = sql.NormalizeAndValidateUsername(args[0]); err != nil {
> 		return errors.Wrap(err, "failed to generate user certificate and key")
> ```

nit, but it's a little odd that the command name and help text refer to it as a "client cert", but in the two error messages you're referring to it as a "user cert"

---

*[pkg/cli/flags.go, line 253 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-Khs2fIucoy_KjgpIgLB:-Khs2fIucoy_KjgpIgLC:bpb33rq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c2423ef3839e978c15ba68b86ddcd990c28754/pkg/cli/flags.go#L253)):*
> ```Go
> 		stringFlag(f, &baseCfg.SSLCertsDir, cliflags.CertsDir, base.DefaultCertsDirectory)
> 
> 		// The start command support the old flags for backwards compatibility.
> ```

Why does only the start command need to continue supporting them? It's odd to continue support for start but not for any of the client commands.

---

*[pkg/cli/flags.go, line 280 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-Khs29tTJui6DPAjFC_2:-Khs29tTJui6DPAjFC_3:b-q7jltz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c2423ef3839e978c15ba68b86ddcd990c28754/pkg/cli/flags.go#L280)):*
> ```Go
> 		f := cmd.Flags()
> 		// CA certificates have a longer expiration time.
> 		durationFlag(f, &certificateLifetime, cliflags.CertificateLifetime, defaultCALifetime)
> ```

You tested that using the same flag object and package-level variable works properly for this and client cert commands below, right? I *think* the pflag library only behaves poorly for environment variables, but I'm not positive off the top of my head and this looks risky given our experience.

---

*[pkg/cli/start.go, line 100 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-Khs34Y_lLCIECgojs9G:-Khs34Y_lLCIECgojs9H:b-otfixx) ([raw file](https://github.com/cockroachdb/cockroach/blob/00185f3df8fc170719a87083632e820891e47264/pkg/cli/start.go#L100)):*
> ```Go
> }
> 
> func initInsecureServer() error {
> ```

We should make a forum announcement about this change in addition to a big warning in the release notes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-:-Khs8I6xsx09tTWp7CcF:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: all files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[pkg/base/config.go, line 130 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-Khryyu7RMjj3zZPZEQZ:-KhsC8DaBd1S0IfGtz8u:b-t0wvoa) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c2423ef3839e978c15ba68b86ddcd990c28754/pkg/base/config.go#L130)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

nit, but you call it `CertFlags` in the environment variable and `CertsFlags` here. You should use the same pluralization everywhere
</blockquote></details>

oops. Sorry, I did that a few times and must have missed this one. Fixed to plural (to match `CertsDir`)

---

*[pkg/base/config.go, line 139 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-KhrjWNrhlkIKJYiIakK:-KhsCYwB9GNnUNnEpwJS:b-6em9hy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c2423ef3839e978c15ba68b86ddcd990c28754/pkg/base/config.go#L139)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

They're deprecated now and will soon be removed. But I'm still not sure we want to be so aggressive about removing them. What harm does it do to keep them around? (I recognize that with 1.0 coming up we either remove them quickly or we'll be stuck with them forever)
</blockquote></details>

Fixed.
My reasoning is pretty much your parenthesis. I'm happy leaving them for compatibility longer (we can even hide the flags from the help message with `flag.Hidden = true`) but at some point I think we need to drop them entirely.

---

*[pkg/cli/cert.go, line 33 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-KhrlGCd2kpgaH0s4bMR:-Khs8ZnZV28Owy8Wti-R:b-u0o0pe) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c2423ef3839e978c15ba68b86ddcd990c28754/pkg/cli/cert.go#L33)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Also, I'd again request that we set these defaults higher such that only people who opt in to rotation ever have to deal with it. 500 years / 100 years, perhaps? Or no expiration?
</blockquote></details>

@bdarnell: doh! Changed to 365. I don't think trying to adjust for leap years helps, this is precisely the reason time.Duration doesn't support days, months, or years which is rather annoying. The flags look odd enough in hours without being unexpected multiples.

@a-robinson: I'm happy to set them to just about anything but I think some type of medium limit + monitoring will be reasonable. In the medium-term, we should also look into integration with PKI solutions. I do realize these are not ideal for people who don't care, and definitely not for more security-conscious folks, but I don't think practically-unlimited values are helpful either.

---

*[pkg/cli/cert.go, line 46 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-KhrmqJ-lgxaFyMQUDah:-KhsDNYs6nesA51yNNQN:b-rosbvl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c2423ef3839e978c15ba68b86ddcd990c28754/pkg/cli/cert.go#L46)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

And if the CA key exists, what happens? Do we generate a new cert with the old key? (this would defeat the "select the ca cert based on which works with the key) approach mentioned below)
</blockquote></details>

We use it if it exists. This would allow people who don't care about rotating the key to just generate a longer-lived CA, and it would also allow externally-generated keys.
Similar to the `--force` flag you mentioned above, we could add a `--allow-key-reuse` flag, but I'm not sure this is strictly necessary.

---

*[pkg/cli/cert.go, line 71 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-KhrmPYfbbKq77MtcMoL:-Khs9_uLHWBnDIgQckLG:b7jqfbj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c2423ef3839e978c15ba68b86ddcd990c28754/pkg/cli/cert.go#L71)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Maybe we should require a `--force` flag to overwrite existing files.
</blockquote></details>

We can, but since we're not plainly overwriting the important ones (`ca.{crt,key}`), mistakes on files that aren't source-controlled are reversible (granted, not trivially). I'm happy to add one if you insist.

---

*[pkg/cli/cert.go, line 84 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-KhrmZYmL2F8Mz4kS3TC:-Khs9uSwjlVjuV1dUj-C:bck5v1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c2423ef3839e978c15ba68b86ddcd990c28754/pkg/cli/cert.go#L84)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Some crypto libraries support simply concatenating the certificate to the private key file. If Go's does too, then we could store a second copy of the cert in the ca.key file and then use that one to avoid ambiguity.
</blockquote></details>

The Go tls library barfs on multi-type PEM files. see `tls.X509KeyPair` in https://golang.org/src/crypto/tls/tls.go?s=5915:5986, L204 (skips non-cert blocks in certificate files) and L230 (fails on certificate blocks in a key file).
Since I've already had to copy/paste some of the tls code, we could go full head-under-water and do the entire pem file handling ourselves, but that has its own problems (as we've covered in my early RFC drafts)

---

*[pkg/cli/cert.go, line 124 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-Khs1PFLjjnGyMCoBpGK:-KhsAv_OG-EQjzre-0FR:b-d4ts9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c2423ef3839e978c15ba68b86ddcd990c28754/pkg/cli/cert.go#L124)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

nit, but it's a little odd that the command name and help text refer to it as a "client cert", but in the two error messages you're referring to it as a "user cert"
</blockquote></details>

fair point. Changing to `client certificate` here and a few lines below. `pkg/security` seems fine.

---

*[pkg/cli/flags.go, line 253 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-Khs2fIucoy_KjgpIgLB:-KhsBIaDirMQNaDWFP-6:bypaccc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c2423ef3839e978c15ba68b86ddcd990c28754/pkg/cli/flags.go#L253)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Why does only the start command need to continue supporting them? It's odd to continue support for start but not for any of the client commands.
</blockquote></details>

Because that's the only thing that would break unexpectedly (and potentially dangerously) if we change flags without much warning.
A few of the client commands could be impacted as well (especially `quit`), but that feels much less risky.

---

*[pkg/cli/flags.go, line 280 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-Khs29tTJui6DPAjFC_2:-KhsBe8DBznQ_fOyhkDa:byno13z) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c2423ef3839e978c15ba68b86ddcd990c28754/pkg/cli/flags.go#L280)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

You tested that using the same flag object and package-level variable works properly for this and client cert commands below, right? I *think* the pflag library only behaves poorly for environment variables, but I'm not positive off the top of my head and this looks risky given our experience.
</blockquote></details>

yup, other commands fail on non-existent flag. The ENV wrapper we wrote is the issue, not pflag.

---

*[pkg/cli/start.go, line 100 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-Khs34Y_lLCIECgojs9G:-KhsBphMphpeqz-XGYaA:bhuulhk) ([raw file](https://github.com/cockroachdb/cockroach/blob/00185f3df8fc170719a87083632e820891e47264/pkg/cli/start.go#L100)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

We should make a forum announcement about this change in addition to a big warning in the release notes.
</blockquote></details>

I agree. I expect to have a forum announcement, a bit warning in the beta notes, and the docs fixed.

---

*[pkg/cli/interactive_tests/test_old_secure.tcl, line 17 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-KhrndHsuAOXr55EZ-lS:-KhsBx3fq6H_0MxPa-YY:bnjzbj1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0c2423ef3839e978c15ba68b86ddcd990c28754/pkg/cli/interactive_tests/test_old_secure.tcl#L17)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We handle `kill` (with the default `SIGTERM`) and `cockroach quit` the same way, so you can just kill the process instead of using `kill -CONT` to see whether it exists and then using `quit`. 
</blockquote></details>

I grabbed this from `test_secure.tcl` so I'm not sure about the reason behind this. This is supposed to be as long-lived as the old flags themselves but I can fix it here and in the others if you like.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703)*
<!-- Sent from Reviewable.io -->

mberhault commented:
I've added `--overwrite [default=false]` and `--allow-ca-key-reuse [default=false]` on the cert commands (the latter only on `create-ca`).
This disallows overwriting files by default, and fails if the CA key exists in `create-ca`. It can still be toggled to allow externally-generated keys or intentional key reuse.

mberhault commented:
Squashing the commit, I left them this way mostly for review, but they can't be reverted individually as some tests break in-between.

I'll wait a bit before merge to make sure my review responses are adequate.

a-robinson commented:
Most recent changes LGTM. I still don't like the cert length, but I need to do some research some time to learn from other systems' experience before I can really make an argument.

---

Reviewed 3 of 3 files at r17.
Review status: 19 of 28 files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-:-KhwL1ytYO5U29T5Qk1Q:bf2sz2)*
<!-- Sent from Reviewable.io -->

mberhault commented:
I'm with you on this part, I don't have good enough data either.
Since these are two constants, I've filed https://github.com/cockroachdb/cockroach/issues/14985 to argue about it. it's also targeted for 1.0 as I would rather avoid these kinds of changes (especially if increasing the lifetime later).

---

Review status: 19 of 28 files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-:-KhwNGMotHEu6vmQgXcj:bqb5dg3)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 3 files at r17, 9 of 10 files at r18.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[pkg/cli/range.go, line 35 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-KhwiVwMrMEhtFfTV4QG:-KhwiVwMrMEhtFfTV4QH:b5xg68s) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae6e3229f9dd6f238ed1fa25a5448af5068aac2c/pkg/cli/range.go#L35)):*
> ```Go
> // MakeDBClient creates a kv client for use in cli tools.
> func MakeDBClient() (*client.DB, *stop.Stopper, error) {
> 	// TODO(marc): Switch back to default user (security.RootUser) when fixed:
> ```

Why did you remove this TODO? Don't we still want to switch this back to RootUser at some point?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-:-Khwity51pCbr4ayqwjr:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[pkg/cli/range.go, line 35 at r15](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703#-KhwiVwMrMEhtFfTV4QG:-KhwjLETEwD5e4LRFv_p:b28nhny) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae6e3229f9dd6f238ed1fa25a5448af5068aac2c/pkg/cli/range.go#L35)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why did you remove this TODO? Don't we still want to switch this back to RootUser at some point?
</blockquote></details>

This is the DB client, the server is hard-coded to only accept `NodeUser` so it's legit.
The other TODO was on `StatusClient` which was indeed a bug which is now fixed and that TODO addressed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14703)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Possibly worth calling out that this requires a decent-sized refresh of https://www.cockroachlabs.com/docs/create-security-certificates.html and https://www.cockroachlabs.com/docs/secure-a-cluster.html.

mberhault commented:
yup, we have https://github.com/cockroachdb/docs/issues/1276 open for that.

jseldess commented:
Docs updated with https://github.com/cockroachdb/docs/pull/1306
